In [64]:
import pandas as pd
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from collections import Counter


df = pd.read_csv('test.csv')


In [65]:
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0  1461          20       RH         80.0    11622   Pave   NaN      Reg   
1  1462          20       RL         81.0    14267   Pave   NaN      IR1   
2  1463          60       RL         74.0    13830   Pave   NaN      IR1   
3  1464          60       RL         78.0     9978   Pave   NaN      IR1   
4  1465         120       RL         43.0     5005   Pave   NaN      IR1   

  LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence MiscFeature  \
0         Lvl    AllPub  ...         120        0    NaN  MnPrv         NaN   
1         Lvl    AllPub  ...           0        0    NaN    NaN        Gar2   
2         Lvl    AllPub  ...           0        0    NaN  MnPrv         NaN   
3         Lvl    AllPub  ...           0        0    NaN    NaN         NaN   
4         HLS    AllPub  ...         144        0    NaN    NaN         NaN   

  MiscVal MoSold  YrSold  SaleType  SaleCondition  
0       0      6    2010        WD         Normal  
1   12500      6    2010        WD         Normal  
2       0      3    2010        WD         Normal  
3       0      6    2010        WD         Normal  
4       0      1    2010        WD         Normal  

[5 rows x 80 columns]

In [66]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1459 entries, 0 to 1458
Data columns (total 80 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1459 non-null   int64  
 1   MSSubClass     1459 non-null   int64  
 2   MSZoning       1455 non-null   object 
 3   LotFrontage    1232 non-null   float64
 4   LotArea        1459 non-null   int64  
 5   Street         1459 non-null   object 
 6   Alley          107 non-null    object 
 7   LotShape       1459 non-null   object 
 8   LandContour    1459 non-null   object 
 9   Utilities      1457 non-null   object 
 10  LotConfig      1459 non-null   object 
 11  LandSlope      1459 non-null   object 
 12  Neighborhood   1459 non-null   object 
 13  Condition1     1459 non-null   object 
 14  Condition2     1459 non-null   object 
 15  BldgType       1459 non-null   object 
 16  HouseStyle     1459 non-null   object 
 17  OverallQual    1459 non-null   int64  
 18  OverallC

This DataFrame consists of 1459 entries and 80 columns, with each row representing a specific instance of data and each column representing a particular attribute or feature. The 'Non-Null Count' column reveals the number of non-null values present in each attribute, indicating potential data completeness. However, several columns contain a considerable number of null values, suggesting potential data incompleteness or missing information. For example, columns like 'Alley', 'PoolQC', 'Fence', and 'MiscFeature' have a relatively low count of non-null values, implying a significant amount of missing data in those columns. Additionally, attributes like 'GarageType', 'GarageYrBlt', 'GarageFinish', 'GarageQual', and 'GarageCond' also have a notable number of null values. This DataFrame likely represents housing-related data, encompassing various features such as lot dimensions, building materials, and amenities. The presence of numerical features like 'LotFrontage', 'LotArea', and 'YearBuilt', alongside categorical variables like 'MSZoning', 'Street', and 'Utilities', suggests a diverse range of information pertaining to properties.

Let's see some descriptive statistics for the data set.

In [67]:
df.describe()

Id   MSSubClass  LotFrontage       LotArea  OverallQual  \
count  1459.000000  1459.000000  1232.000000   1459.000000  1459.000000   
mean   2190.000000    57.378341    68.580357   9819.161069     6.078821   
std     421.321334    42.746880    22.376841   4955.517327     1.436812   
min    1461.000000    20.000000    21.000000   1470.000000     1.000000   
25%    1825.500000    20.000000    58.000000   7391.000000     5.000000   
50%    2190.000000    50.000000    67.000000   9399.000000     6.000000   
75%    2554.500000    70.000000    80.000000  11517.500000     7.000000   
max    2919.000000   190.000000   200.000000  56600.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1459.000000  1459.000000   1459.000000  1444.000000  1458.000000  ...   
mean      5.553804  1971.357779   1983.662783   100.709141   439.203704  ...   
std       1.113740    30.390071     21.130467   177.625900   455.268042  ...   
min       1.000000  1879.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1953.000000   1963.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1992.000000     0.000000   350.500000  ...   
75%       6.000000  2001.000000   2004.000000   164.000000   753.500000  ...   
max       9.000000  2010.000000   2010.000000  1290.000000  4010.000000  ...   

        GarageArea   WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  \
count  1458.000000  1459.000000  1459.000000    1459.000000  1459.000000   
mean    472.768861    93.174777    48.313914      24.243317     1.794380   
std     217.048611   127.744882    68.883364      67.227765    20.207842   
min       0.000000     0.000000     0.000000       0.000000     0.000000   
25%     318.000000     0.000000     0.000000       0.000000     0.000000   
50%     480.000000     0.000000    28.000000       0.000000     0.000000   
75%     576.000000   168.000000    72.000000       0.000000     0.000000   
max    1488.000000  1424.000000   742.000000    1012.000000   360.000000   

       ScreenPorch     PoolArea       MiscVal       MoSold       YrSold  
count  1459.000000  1459.000000   1459.000000  1459.000000  1459.000000  
mean     17.064428     1.744345     58.167923     6.104181  2007.769705  
std      56.609763    30.491646    630.806978     2.722432     1.301740  
min       0.000000     0.000000      0.000000     1.000000  2006.000000  
25%       0.000000     0.000000      0.000000     4.000000  2007.000000  
50%       0.000000     0.000000      0.000000     6.000000  2008.000000  
75%       0.000000     0.000000      0.000000     8.000000  2009.000000  
max     576.000000   800.000000  17000.000000    12.000000  2010.000000  

[8 rows x 37 columns]

This table provides descriptive statistics for numeric columns in the DataFrame. Each row corresponds to a statistical measure calculated across all values in the respective column. For instance, 'count' represents the number of non-null values in each column, 'mean' signifies the average value, 'std' denotes the standard deviation, 'min' indicates the minimum value, '25%' represents the first quartile (25th percentile), '50%' is the median (50th percentile), and '75%' signifies the third quartile (75th percentile). The 'max' row displays the maximum value in each column. These statistics offer insights into the central tendency, dispersion, and range of values within each numerical attribute.

The analysis of the provided dataset reveals several insights. There are missing values in some columns, notably 'LotFrontage', suggesting incomplete data. Variability across columns varies, with some having high standard deviations indicating significant variability, while others exhibit less variation. The range of values varies widely across different columns, indicating diverse data distributions. Potential outliers are identified by comparing minimum and maximum values, such as the presence of outliers in the 'PoolArea' column. Temporal trends can be inferred from columns related to years, offering insights into changes over time. Additionally, the presence of zero values in certain columns, despite non-zero means, raises questions about missing data or specific property characteristics. Further exploration through visualization and statistical analysis is recommended to better understand the dataset and make informed decisions for subsequent processing or modeling.

Let's take a look at some zero PoolArea and LotArea.


In [68]:
df_PoolArea_zero = df[df.PoolArea == 0]

df_PoolArea_zero.head(7)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0  1461          20       RH         80.0    11622   Pave   NaN      Reg   
1  1462          20       RL         81.0    14267   Pave   NaN      IR1   
2  1463          60       RL         74.0    13830   Pave   NaN      IR1   
3  1464          60       RL         78.0     9978   Pave   NaN      IR1   
4  1465         120       RL         43.0     5005   Pave   NaN      IR1   
5  1466          60       RL         75.0    10000   Pave   NaN      IR1   
6  1467          20       RL          NaN     7980   Pave   NaN      IR1   

  LandContour Utilities  ... ScreenPorch PoolArea PoolQC  Fence MiscFeature  \
0         Lvl    AllPub  ...         120        0    NaN  MnPrv         NaN   
1         Lvl    AllPub  ...           0        0    NaN    NaN        Gar2   
2         Lvl    AllPub  ...           0        0    NaN  MnPrv         NaN   
3         Lvl    AllPub  ...           0        0    NaN    NaN         NaN   
4         HLS    AllPub  ...         144        0    NaN    NaN         NaN   
5         Lvl    AllPub  ...           0        0    NaN    NaN         NaN   
6         Lvl    AllPub  ...           0        0    NaN  GdPrv        Shed   

  MiscVal MoSold  YrSold  SaleType  SaleCondition  
0       0      6    2010        WD         Normal  
1   12500      6    2010        WD         Normal  
2       0      3    2010        WD         Normal  
3       0      6    2010        WD         Normal  
4       0      1    2010        WD         Normal  
5       0      4    2010        WD         Normal  
6     500      3    2010        WD         Normal  

[7 rows x 80 columns]

In the provided data subset, focusing on the 'PoolArea' column, entries with zero values are examined. Taking the first seven rows as an example, it's apparent that several properties have a 'PoolArea' value of zero, indicating the absence of a pool. For instance, in the first row, property Id 1461 has a 'PoolArea' of zero, suggesting no pool in that property. This pattern persists across multiple rows, reinforcing the notion that zero values in the 'PoolArea' column likely signify the absence of pools in the respective properties. Further investigation could involve cross-referencing this data with additional property features or external sources to validate these assumptions.



In [69]:
df_PoolArea_zero.count()

Id               1453
MSSubClass       1453
MSZoning         1449
LotFrontage      1226
LotArea          1453
                 ... 
MiscVal          1453
MoSold           1453
YrSold           1453
SaleType         1452
SaleCondition    1453
Length: 80, dtype: int64

The output above indicates the count of non-null values for each column in the DataFrame where the 'PoolArea' column has a value of zero. It suggests that out of 1459 entries, there are 1453 entries where 'PoolArea' is zero. Additionally, it shows the count of non-null values for other columns in the DataFrame, ranging from 1449 to 1459 depending on the column. This information helps in understanding the distribution of non-null values across different columns in the dataset where 'PoolArea' is zero.

In [70]:
df_PoolArea_0count = df.groupby('PoolArea')[['Id']].count()

print(df_PoolArea_0count)

            Id
PoolArea      
0         1453
144          1
228          1
368          1
444          1
561          1
800          1


The output above indicates the count of entries grouped by the 'PoolArea' column. It shows that there are 1453 entries with a 'PoolArea' value of zero, while the rest have different values ranging from 144 to 800. This analysis helps in understanding the distribution of entries based on the presence or absence of pool areas, with the majority having no pool area.

I count the zero value in the PoolArea cloumn and there are 1453 rows in zero value. In case I drop too many raw data to keep the data integrity, I decide to retain these rows and replace zero values with null values.

So does the LotArea column.

In [71]:
df_LotArea_0count = df.groupby('LotArea')[['Id']].count()
print(df_LotArea_0count)

         Id
LotArea    
1470      1
1476      1
1477      1
1484      1
1488      1
...      ..
47007     1
47280     1
50102     1
51974     1
56600     1

[1106 rows x 1 columns]


This output indicates the count of occurrences for each unique value in the 'LotArea' column of the DataFrame. Each row represents a unique lot area value along with the count of how many times that value appears in the dataset. For example, there is one occurrence of a lot area of 1470 square units, one occurrence of 1476 square units, and so on. The total number of unique lot area values is 1106.

It's just has a small number of zero value rows in runtime column, so I decide to drop them.

Finally, let's investigate the ScreenPorch column to decide whether drop zero or just replace it with null value

In [72]:
df_ScreenPorch_0count = df.groupby('ScreenPorch')[['Id']].count()
print(df_ScreenPorch_0count)

               Id
ScreenPorch      
0            1319
64              1
84              1
88              1
92              2
...           ...
322             1
342             1
348             1
490             1
576             1

[75 rows x 1 columns]


This output presents the count of occurrences for each unique value in the 'ScreenPorch' column of the DataFrame. Each row represents a unique screen porch area value along with the count of how many times that value appears in the dataset. For instance, there are 1319 instances where the screen porch area is 0 square units, one occurrence of 64 square units, one of 84 square units, and so forth. In total, there are 75 unique screen porch area values.

I count the zero value in the PoolArea cloumn and there are 1319 rows in zero value. In case I drop too many raw data to keep the data integrity, I decide to retain these rows and replace zero values with null values.

In [ ]:
# quartile function
def cut_into_quantile(dfname ,column_name):
# find quartile, max and min values
    min_value = dfname[column_name].min()
    first_quantile = dfname[column_name].describe()[4]
    second_quantile = dfname[column_name].describe()[5]
    third_quantile = dfname[column_name].describe()[6]
    max_value = dfname[column_name].max()
# Bin edges that will be used to "cut" the data into groups
    bin_edges = [ min_value, first_quantile, second_quantile, third_quantile, max_value]
# Labels for the four budget level groups
    bin_names = [ 'Low', 'Medium', 'Moderately High', 'High'] 
# Creates budget_levels column
    name = '{}_levels'.format(column_name)
    dfname[name] = pd.cut(dfname[column_name], bin_edges, labels=bin_names, include_lowest = True)
    return dfname